plugins/bcm57xx/fu-bcm57xx-recovery-device.c

```diff
@@ -0,0 +1,753 @@
+						  REG_NVM_ACCESS, &tmp.r32, error))
+		return FALSE;
+	tmp.bits.Enable = TRUE;
+	tmp.bits.WriteEnable = TRUE;
+	return fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						     REG_NVM_ACCESS, tmp.r32, error);
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_acquire_lock (FuBcm57xxRecoveryDevice *self,
+					       GError **error)
+{
+	BcmRegNVMSoftwareArbitration tmp = { 0 };
+	g_autoptr(GTimer) timer = g_timer_new ();
+	tmp.bits.ReqSet1 = 1;
+	if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						   REG_NVM_SOFTWARE_ARBITRATION,
+						   tmp.r32, error))
+		return FALSE;
+	do {
+		if (!fu_bcm57xx_recovery_device_bar_read (self,
+							  FU_BCM57XX_BAR_DEVICE,
+							  REG_NVM_SOFTWARE_ARBITRATION,
+							  &tmp.r32, error))
+			return FALSE;
+		if (tmp.bits.ArbWon1)
+			return TRUE;
+		if (g_timer_elapsed (timer, NULL) > 0.2)
+			break;
+	} while (TRUE);
+
+	/* timed out */
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_TIMED_OUT,
+			     "timed out trying to acquire lock #1");
+	return FALSE;
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_release_lock (FuBcm57xxRecoveryDevice *self,
+					       GError **error)
+{
+	BcmRegNVMSoftwareArbitration tmp = { 0 };
+	tmp.r32 = 0;
+	tmp.bits.ReqClr1 = 1;
+	return fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						     REG_NVM_SOFTWARE_ARBITRATION, tmp.r32,
+						     error);
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_wait_done (FuBcm57xxRecoveryDevice *self, GError **error)
+{
+	BcmRegNVMCommand tmp = { 0 };
+	g_autoptr(GTimer) timer = g_timer_new ();
+	do {
+		if (!fu_bcm57xx_recovery_device_bar_read (self,
+							  FU_BCM57XX_BAR_DEVICE,
+							  REG_NVM_COMMAND, &tmp.r32,
+							  error))
+			return FALSE;
+		if (tmp.bits.Done)
+			return TRUE;
+		if (g_timer_elapsed (timer, NULL) > 0.2)
+			break;
+	} while (TRUE);
+
+	/* timed out */
+	g_set_error_literal (error,
+			     G_IO_ERROR,
+			     G_IO_ERROR_TIMED_OUT,
+			     "timed out");
+	return FALSE;
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_clear_done (FuBcm57xxRecoveryDevice *self, GError **error)
+{
+	BcmRegNVMCommand tmp = { 0 };
+	tmp.bits.Done = 1;
+	return fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						 REG_NVM_COMMAND, tmp.r32, error);
+}
+
+static gboolean
+fu_bcm57xx_recovery_device_nvram_read (FuBcm57xxRecoveryDevice *self,
+				       guint32 address, guint32 *buf, gsize bufsz,
+				       GError **error)
+{
+	for (guint i = 0; i < bufsz; i++) {
+		BcmRegNVMCommand tmp = { 0 };
+		guint32 val32 = 0;
+		if (!fu_bcm57xx_recovery_device_nvram_clear_done (self, error))
+			return FALSE;
+		if (!fu_bcm57xx_recovery_device_bar_write (self, FU_BCM57XX_BAR_DEVICE,
+						       REG_NVM_ADDR, address, error))
+			return FALSE;
+		tmp.bits.Doit = 1;
+		/* write one word at a time to avoid data corruption issue */
+		tmp.bits.First = TRUE;
+		tmp.bits.Last = TRUE;
```

Comment:
So, I've been running into an issue when using a function other than 0 often times out when reading individual words. I don't see this with the utility I have, or when I change this code to do a single burst read.

I do not see this issue when using function 0, only function 1 or later (fwupd is always using the last function, not the first)

Also note that the ethtool apis show the same issue when using a function other than 0. (the tg3 driver does reads similar to this, which results in the timeout).

Reply:
Also note that if I update bcmflash to use this sequence, it also ends up hanging in certain cases.

So, the most reliable way to do this here may be to use something like:
		if (i ==0)
			tmp.bits.First = TRUE;
		else
			tmp.bits.First = FALSE;
		if (i + sizeof(guint32) < bufsz)
			tmp.bits.Last = FALSE;
		else
			tmp.bits.Last = TRUE;

Reply:
So maybe just:

		tmp.bits.First = i == 0;
		tmp.bits.Last = i == bufsz - 1;

`i` is incrementing the address in guint32's, so I think it's just `- 1`